File: src/normalization/parsers.py
import re
import math
from typing import Union

def parse_float(value: Union[str, float, None]) -> float:
    """
    Parses a float from a string or float value, handling common currency symbols and formatting.
    Returns 0.0 if value is None or cannot be parsed.
    """
    if value is None:
        return 0.0
    if isinstance(value, (float, int)):
        return float(value)
    
    # Remove common currency symbols and whitespace
    cleaned_value = str(value).strip().lower()
    cleaned_value = re.sub(r'(?:rs\.?|inr|\$|€|£)', '', cleaned_value).strip()
    # Remove commas
    cleaned_value = cleaned_value.replace(',', '')

    if not cleaned_value:
        return 0.0
        
    # Handle "Billed + Free" formats (e.g. "10+2", "4.50+.50")
    if "+" in cleaned_value:
        try:
            parts = cleaned_value.split('+')
            # Extract the FIRST number found (Billed Qty)
            first_part = parts[0]
            match = re.search(r'\d+(\.\d+)?', first_part)
            if match:
                return float(match.group())
        except:
            pass 
    
    match = re.search(r'-?\d+(\.\d+)?', cleaned_value)
    if match:
        return float(match.group())
    return 0.0

def parse_quantity(value: Union[str, float, None]) -> int:
    """
    Parses a quantity string, handling sums (e.g. '10+2') and rounding UP to nearest integer.
    """
    if value is None:
        return 0
        
    if isinstance(value, (float, int)):
        return math.ceil(value)
        
    cleaned_value = str(value).strip().lower()
    cleaned_value = re.sub(r'(?:rs\.?|inr|\$|€|£)', '', cleaned_value).strip()
    cleaned_value = cleaned_value.replace(',', '')
    
    if not cleaned_value:
        return 0
        
    total_qty = 0.0
    
    # Check for split sums "1.5 + 1.5" or "10+2"
    if "+" in cleaned_value:
        parts = cleaned_value.split('+')
        for part in parts:
            match = re.search(r'-?\d+(\.\d+)?', part.strip())
            if match:
                total_qty += float(match.group())
    else:
        match = re.search(r'-?\d+(\.\d+)?', cleaned_value)
        if match:
            total_qty = float(match.group())
            
    return float(total_qty)

File: src/normalization/parsers.js

/**
 * Parses a float from a string or number, handling currency symbols and formatting.
 * @param {string|number|null} value 
 * @returns {number}
 */
export function parseFloatClean(value) {
    if (value === null || value === undefined) return 0.0;
    if (typeof value === 'number') return value;

    let cleanedValue = String(value).trim().toLowerCase();
    // Remove currency symbols (rs., inr, $, €, £)
    cleanedValue = cleanedValue.replace(/(?:rs\.?|inr|\$|€|£)/g, '').trim();
    // Remove commas
    cleanedValue = cleanedValue.replace(/,/g, '');

    if (!cleanedValue) return 0.0;

    // Handle "Billed + Free" formats (e.g. "10+2")
    if (cleanedValue.includes('+')) {
        try {
            const parts = cleanedValue.split('+');
            const firstPart = parts[0];
            const match = firstPart.match(/\d+(\.\d+)?/);
            if (match) {
                return parseFloat(match[0]);
            }
        } catch (e) {
            // ignore
        }
    }

    const match = cleanedValue.match(/-?\d+(\.\d+)?/);
    if (match) {
        return parseFloat(match[0]);
    }
    return 0.0;
}

/**
 * Parses a quantity string, handling sums (e.g. '10+2') and rounding UP to nearest integer.
 * @param {string|number|null} value 
 * @returns {number}
 */
export function parseQuantity(value) {
    if (value === null || value === undefined) return 0;
    if (typeof value === 'number') return Math.ceil(value);

    let cleanedValue = String(value).trim().toLowerCase();
    cleanedValue = cleanedValue.replace(/(?:rs\.?|inr|\$|€|£)/g, '').trim();
    cleanedValue = cleanedValue.replace(/,/g, '');

    if (!cleanedValue) return 0;

    let totalQty = 0.0;

    if (cleanedValue.includes('+')) {
        const parts = cleanedValue.split('+');
        for (const part of parts) {
            const match = part.trim().match(/-?\d+(\.\d+)?/);
            if (match) {
                totalQty += parseFloat(match[0]);
            }
        }
    } else {
        const match = cleanedValue.match(/-?\d+(\.\d+)?/);
        if (match) {
            totalQty = parseFloat(match[0]);
        }
    }

    // Python logic was returning float(total_qty) but the docstring said rounding UP. 
    // Wait, Python code: return float(total_qty). 
    // The parser logic says "rounding UP to nearest integer" but implementation returns float.
    // I will stick to what the Python implementation DOES, which is returning the float sum.
    // However, the Python function `parse_quantity` line 49 `if isinstance... return math.ceil` implies integer intent.
    // But line 72 `return float(total_qty)` returns float.
    // I will align with "return float" for now to match behavior.
    return totalQty;
}
